pkg/handler/collector/oci/oci.go

```diff
@@ -54,84 +54,87 @@ func NewOCICollector(ctx context.Context, repoTags map[string][]string, poll boo
-	if o.poll {
-		for {
-			for repo, tags := range o.repoTags {
-				// when polling if tags are specified, it will never get any new tags
-				// that might be added after the fact. Defeating the point of the polling
-				if len(tags) > 0 {
-					return errors.New("image tag should not specified when using polling")
-				}
-				err := o.getTagsAndFetch(ctx, repo, tags, docChannel)
-				if err != nil {
-					return err
-				}
-				// set interval to about 5 mins or more
-				time.Sleep(o.interval)
-			}
-		}
-	} else {
+	if docChannel == nil {
+		return fmt.Errorf("nil channel passed to %s", OCICollector)
+	}
+
+	for {
 		for repo, tags := range o.repoTags {
-			err := o.getTagsAndFetch(ctx, repo, tags, docChannel)
-			if err != nil {
-				return err
+			if len(tags) > 0 && o.poll {
+				return errors.New("image tag should not specified when using polling")
+			}
+			if err := o.getTagsAndFetch(ctx, repo, tags, docChannel); err != nil {
+				return fmt.Errorf("fetching artifacts for %s: %w", repo, err)
 			}
 		}
+		if !o.poll {
+			break
+		}
+		select {
+		case <-ctx.Done():
+			// This way the function will not block for a fixed
+			// interval and can be stopped immediately when the context is canceled.
+			return ctx.Err()
+		case <-time.After(o.interval):
+		}
 	}
 
 	return nil
 }
 
 func (o *ociCollector) getTagsAndFetch(ctx context.Context, repo string, tags []string, docChannel chan<- *processor.Document) error {
-	rcOpts := []regclient.Opt{}
-	rcOpts = append(rcOpts, regclient.WithDockerCreds())
-	rcOpts = append(rcOpts, regclient.WithDockerCerts())
+	if repo == "" {
+		return errors.New("repository name not specified")
+	}
 
-	if len(tags) > 0 {
-		for _, tag := range tags {
-			if tag == "" {
-				return errors.New("image tag not specified to fetch")
-			}
-			imageTag := fmt.Sprintf("%v:%v", repo, tag)
-			r, err := ref.New(imageTag)
-			if err != nil {
-				return err
-			}
+	if docChannel == nil {
+		return errors.New("invalid document channel")
+	}
 
-			rc := regclient.New(rcOpts...)
-			defer rc.Close(ctx, r)
+	rcOpts := []regclient.Opt{
+		regclient.WithDockerCreds(),
+		regclient.WithDockerCerts(),
+	}
 
-			err = o.fetchOCIArtifacts(ctx, repo, rc, r, docChannel)
-			if err != nil {
-				return err
-			}
-		}
+	rc := regclient.New(rcOpts...)
+
+	repoRef, err := ref.New(repo)
+	defer rc.Close(ctx, repoRef)
+
+	if err != nil {
+		return fmt.Errorf("parsing repository reference: %w", err)
+	}
+
+	var fetchTags []string
+	if len(tags) > 0 {
+		fetchTags = tags
 	} else {
-		r, err := ref.New(repo)
+		allTags, err := rc.TagList(ctx, repoRef)
 		if err != nil {
-			return err
+			return fmt.Errorf("reading tags for %s: %w", repo, err)
 		}
 
-		rc := regclient.New(rcOpts...)
-		defer rc.Close(ctx, r)
+		// Filter out tags that are not images
+		for _, tag := range allTags.Tags {
+			// filter out tags that looking for sha256- followed by a 64 character hex string
+			if !(strings.HasPrefix(tag, "sha256-") && len(tag) == 71) {
```

Comment:
Thanks @sudo-bmitch! This should also be done later to remove the suffix list and look for sha256- followed by a 64 character hex string at line 176.

Reply:
OK, Which is the tag in `176`? Also will the tag check replace this `if !contains(o.checkedDigest[repo], digestTag) {`?

It isn't clear to me.

Reply:
@pxp928  Friendly ping.

Reply:
Could we merge this and address it in the upcoming PR? In my opinion, this is not a critical issue, as it is an existing implementation. 

It simplifies the process for me.

Reply:
sure its fine for now, until we move on to the `OCI referrers` which will remove the suffix and prefix check anyways.